### PR TITLE
1159303 - delete agent queues on consumer unregistration.

### PR DESCRIPTION
**Pull request by jortel**
https://bugzilla.redhat.com/show_bug.cgi?id=1159303

The NotFound exception is new in gofer 2.5 which contains support for amqp 1.0 (proton).  We can delay merging until that work is complete and gofer 2.5 is released.

### CODE REVIEW

File: server/pulp/server/managers/consumer/agent.py [hunk 1]
```diff
@@ -14,10 +14,13 @@
 """
 
 import sys
+
 from logging import getLogger
 from uuid import uuid4
+from gettext import gettext as _
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: should come before logging import to be alphabetical

File: server/pulp/server/managers/consumer/agent.py [hunk 2]
```diff
@@ -30,6 +33,12 @@
 from pulp.server.managers import factory as managers
 from pulp.server.async.task_status_manager import TaskStatusManager
 from pulp.server.agent import Context
+from pulp.server.async.tasks import Task
+
+
+QUEUE_DELETE_DELAY = 600  # 10 min.
+QUEUE_DELETE_FAILED = _('queue %(name)s cannot be deleted: %(reason)s')
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: nice internationalization

File: server/pulp/server/managers/consumer/agent.py [hunk 3]
```diff
@@ -30,6 +33,12 @@
 from pulp.server.managers import factory as managers
 from pulp.server.async.task_status_manager import TaskStatusManager
 from pulp.server.agent import Context
+from pulp.server.async.tasks import Task
+
+
+QUEUE_DELETE_DELAY = 600  # 10 min.
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: Do we want this configurable? I'm not sure myself but I want to as the question.

File: server/pulp/server/managers/consumer/agent.py [hunk 4]
```diff
@@ -404,6 +418,46 @@ def _unbindings(bindings):
...
+            manager.get_consumer(consumer_id)
+            return  # still registered (abort)
+        except MissingResource:
+            # expected
+            pass
+        agent = PulpAgent()
+        agent.delete_queue(url, name)
+
+
+@task(base=Task)
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: This will cause the task to be recorded with a task status. That is what you want right? Can you verify that the task_status records that get created look right? If you want it to occur but not be user facing then you can not use `base=Task`

File: server/pulp/server/managers/consumer/agent.py [hunk 5]
```diff
@@ -404,6 +418,46 @@ def _unbindings(bindings):
...
+        Delete the agent queue.
+        :param url: The broker URL.
+        :type url: str
+        :param name: The queue name.
+        :type name: str
+        :param consumer_id: The consumer ID.
+        :type consumer_id: str
+        """
+        try:
+            manager = managers.consumer_manager()
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: Could this statement raise the MissingResource exception? If not it should be moved outside the try/except block

File: server/pulp/server/agent/direct/pulpagent.py [hunk 6]
```diff
@@ -76,6 +77,23 @@ def cancel(context, task_id):
...
+    def delete_queue(url, name):
+        """
+        Purge and delete the agent queue.
+        :param url: The broker URL.
+        :type url: str
+        :param name: The queue name.
+        :type name: str
+        """
+        try:
+            queue = Queue(name)
```
Comment (bmbouter) [inline server/pulp/server/agent/direct/pulpagent.py]: Which of these statements can raise NotFound? The others should be moved outside of the try/except block.

File: server/pulp/server/managers/consumer/agent.py [hunk 7]
```diff
@@ -404,6 +418,46 @@ def _unbindings(bindings):
             agent_bindings.append(agent_binding)
         return agent_bindings
 
+    @staticmethod
+    def delete_queue(url, name, consumer_id):
+        """
+        Delete the agent queue.
+        :param url: The broker URL.
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: I usually put a whitespace line between the description and the parameters for readability.

File: server/pulp/server/managers/consumer/agent.py [hunk 8]
```diff
@@ -404,6 +418,46 @@ def _unbindings(bindings):
...
+    :param url: The broker URL.
+    :type url: str
+    :param name: The queue name.
+    :type name: str
+    :param consumer_id: The consumer ID.
+    :type consumer_id: str
+    """
+    try:
+        AgentManager.delete_queue(url, name, consumer_id)
+        logger.info(QUEUE_DELETED, {'name': name})
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: This probably isn't a line you are guarding for exceptions. This should be moved to an 'else' clause so this would have try/except/else as codepath branches.

File: server/pulp/server/managers/consumer/agent.py [hunk 9]
```diff
@@ -416,6 +470,7 @@ def __init__(self, units):
         :param units: A list of content units.
         :type units: list
         """
+        super(Units, self).__init__()
```
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: Why is this needed? How does this impact the code behaviors?

File: server/test/unit/server/agent/direct/test_agent.py [hunk 10]
```diff
@@ -28,8 +28,44 @@ def test_capabilities(self):
...
+
+    @patch('pulp.server.agent.direct.pulpagent.Queue')
+    @patch('pulp.server.agent.direct.pulpagent.NotFound', NotFound)
+    def test_delete_queue_not_found(self, queue):
+        url = 'test-url'
+        name = 'test-queue'
+        queue.return_value.purge.side_effect = NotFound
+
+        # test
+        agent = PulpAgent()
```
Comment (bmbouter) [inline server/test/unit/server/agent/direct/test_agent.py]: If the code is broken then an exception will be raised here. consider putting it in a try/except block and in the except block putting a self.fail(). That will cause it to be marked as a failure which is more accurate than an Error.

File: server/test/unit/server/managers/consumer/test_agent_manager.py [hunk 11]
```diff
@@ -536,6 +543,75 @@ def test_get_distributor():
...
+    @patch('pulp.server.managers.consumer.agent.managers.consumer_manager')
+    @patch('pulp.server.managers.consumer.agent.PulpAgent')
+    def test_delete_queue(self, agent, consumer_manager):
+        url = 'test-url'
+        queue = 'test-queue'
+        consumer_id = 'test-consumer'
+        consumer_manager.return_value.get_consumer.side_effect = MissingResource
+
+        # test
+        AgentManager.delete_queue(url, queue, consumer_id)
```
Comment (bmbouter) [inline server/test/unit/server/managers/consumer/test_agent_manager.py]: Similar try/except suggested here as [this comment](https://github.com/pulp/pulp/pull/1583/files#r23771860).

**Comment by bmbouter** [comment]
flake8 all pass, tests all pass for me, and all new modules have 100% code coverage! Good job. Implementing the suggestions before merging would be good. Besides code cleanup there are questions regarding the task_status data created by the delete_queue task and configurability of the the queue delete delay.

**Comment by bmbouter** [comment]
Also I see that 2.6-testing was merged in after your commit. For working branches the best is to rebase your commit onto 2.6-testing again so that your commit will come after HEAD on 2.6-testing. We had a lot of challenges in the Django branching stuff because we were mixing merge and rebase. We should have used rebase, and so I'm passing along this info. Ask me if its not clear why rebase is better than merge.

**Comment by bmbouter** [comment]
The 'jortel-' in the branch name is redundant since it shows the PR from your fork already. Then it reads as 'jortel:jortel-1159303'

File: server/pulp/server/managers/consumer/agent.py (hunk 1 above)
Comment (jortel) [inline server/pulp/server/managers/consumer/agent.py]: alphabetical?  I'm not aware of this requirement in pep8.  can you point me to the text?

File: server/pulp/server/managers/consumer/agent.py (hunk 3 above)
Comment (jortel) [inline server/pulp/server/managers/consumer/agent.py]: I'm not sure either but leaning against.  Figured that when in doubt, error on the side of less complexity.

File: server/pulp/server/managers/consumer/agent.py (hunk 4 above)
Comment (jortel) [inline server/pulp/server/managers/consumer/agent.py]: Good catch.  Nope, I do not want this recorded in our task status collection.

File: server/pulp/server/managers/consumer/agent.py (hunk 9 above)
Comment (jortel) [inline server/pulp/server/managers/consumer/agent.py]: when overriding constructors, the superclass constructor should always be called first unless the explicit intent is to completely replace its behavior (and done with extreme caution).  Not calling the superclass constructor can result in partially initialized objects.

File: server/pulp/server/managers/consumer/agent.py (hunk 1 above)
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: It's not a PEP8 requirement, just a healthy convention. Notice that almost all imports are alphabetized in large portions of the codebase.

File: server/pulp/server/managers/consumer/agent.py (hunk 3 above)
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: I think this is fine for now. The issue was first noticed by QE so in terms of tests it may be too long. This is already a huge improvement so lets leave it alone for now.

File: server/pulp/server/managers/consumer/agent.py (hunk 9 above)
Comment (bmbouter) [inline server/pulp/server/managers/consumer/agent.py]: Agreed. This change is more correct so leave it. I just hope it doesn't cause an unexpected change.

File: server/pulp/server/managers/consumer/agent.py (hunk 1 above)
Comment (rbarlow) [inline server/pulp/server/managers/consumer/agent.py]: On 01/29/2015 10:29 AM, Jeff Ortel wrote:
> alphabetical? I'm not aware of this requirement in pep8. can you point
> me to the text?

Agreed with what @bmbouter said, and I'll add that it's specifically
helpful because it groups imports from the same module and nearby
modules together.

**Comment by jortel** [comment]
ok test

**Comment by jortel** [comment]
ok test

**Comment by beav** [comment]
ok test
